python/ray/dag/compiled_dag_node.py

```diff
@@ -496,6 +603,145 @@ def exec_operation(
+        """
+        self._stream_buffer[op.next_operation()] = data
+
+    def reset_stream_buffer(self, op: _DAGNodeOperation) -> Any:
+        """
+        Retrieve the intermediate result of a READ or COMPUTE operation
+        from the stream buffer and clear the entry from the buffer.
+
+        Returns:
+            The intermediate result of a READ or COMPUTE operation.
+        """
+        return self._stream_buffer.pop(op)
+
+    def _stream_read(self, op: _DAGNodeOperation) -> bool:
+        """
+        Read input data from upstream DAG nodes and cache the intermediate result.
+
+        Args:
+            op: The READ operation.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        exit = False
+        try:
+            input_data = self.input_reader.read()
+            self.set_stream_buffer(op, input_data)
+        except RayChannelError:
+            # Channel closed. Exit the loop.
+            exit = True
+        return exit
+
+    def _stream_compute(self, op: _DAGNodeOperation, exec_stream, class_handle) -> bool:
+        """
+        Retrieve the intermediate result from the READ operation and perform the
+        computation on the provided execution stream. Then, cache the new
+        intermediate result.
+
+        Args:
+            op: The compute operation.
+            exec_stream: The CUDA stream to execute the compute operation.
+            class_handle: An instance of the class to which the actor belongs. For
+                example, the type of `class_handle` is <class 'xxxx.Worker'> if the
+                actor belongs to the `class Worker` class.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        input_data = self.reset_stream_buffer(op)
+        method = getattr(class_handle, self.method_name)
+        try:
+            _process_return_vals(input_data, return_single_output=False)
+        except Exception as exc:
+            # Previous task raised an application-level exception.
+            # Propagate it and skip the actual task. We don't need to wrap the
+            # exception in a RayTaskError here because it has already been wrapped
+            # by the previous task.
+            self.set_stream_buffer(exc)
+            return False
+
+        channel_results = []
+        for entry in input_data:
+            if isinstance(entry, tuple):
+                channel_result, event = entry
+                if event:
+                    event.synchronize()
+            else:
+                channel_result = entry
+            channel_results.append(channel_result)
+
+        resolved_inputs = []
+        for task_input in self.task_inputs:
+            resolved_inputs.append(task_input.resolve(channel_results))
+
+        import cupy as cp
+
+        exec_event = cp.cuda.Event()
+        with exec_stream:
```

Comment:
It would be best if we can avoid implicitly setting the execution stream (user may have code that manages streams themselves).

Reply:
I think the current behavior is if user uses another stream in their UDF, that will take precedence. Is it a behavior we don't want? Can you elaborate a bit on the problem and the stream-management-contract-with-user you have on mind?

Reply:
Hmm in general I'm wary of implicitly overwriting user defaults but it could be okay in this case since as you said, the user stream will take precedence. Once we have an API to set the execution stream, this also becomes less of an issue and more of a debate of what the better default behavior is (just use the default stream or create one).

One tricky thing is that the CUDA default stream also has different synchronization semantics depending on compilation flags...: https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html

Reply:
I want to avoid using a new stream for execution. It is confusing behavior for users, and if they do something in main thread with a default stream, it can cause issues (tho it is not common).

Reply:
Is Stephanie's original suggestion also to use the passed-in stream as execution stream rather than creating a new one? like this Sang's comment here https://github.com/ray-project/ray/pull/47586#discussion_r1767014682. @stephanie-wang 

If so, I can change to that.

Reply:
I think we should not use any stream for execution (so it uses default from users). And only when it is explicitly passed, we should allow to overwrite (in reality, users may just also do it themselves if required). let's minimze magic behaviors if possible

Reply:
@rkooo567 I will have a quick chat with you tomorrow.